File: ios/Runner/FilePicker.swift
import Flutter
import PhotosUI
import UIKit
import UniformTypeIdentifiers

/// Choosing a file, through the system picker and nothing else.
///
/// The same shape as `android/.../FilePicker.kt` and for the same reason: this
/// asks for **no permission at all**. `UIDocumentPickerViewController` is drawn
/// by the system, the person picks one file, and this process is handed that
/// one file.
///
/// Pictures go through `PHPickerViewController`, which asks for nothing either.
/// It was avoided here on the belief that it cost
/// `NSPhotoLibraryUsageDescription`, and that has not been true since iOS 14:
/// the grid is drawn by a separate process that this application cannot see
/// into, and what comes back is the one picture that was tapped. There is no
/// permission, no prompt, and nothing on the privacy screen.
///
/// Which is why sending a photograph now opens the photographs rather than the
/// file system. Somebody attaching a picture was being shown a list of folders
/// and having to go and find it, which is the wrong question asked politely.
///
/// # Why the bytes are copied here
///
/// A URL from the picker is security-scoped and valid inside
/// `startAccessingSecurityScopedResource`. Handing it to Dart to open later
/// produces a permission failure at whatever moment the person finally presses
/// send, which is the worst place to discover it.
class FilePicker: NSObject, UIDocumentPickerDelegate {

    static let channel = "rotelyx/files"

    /// Refused before it is read, so a huge file is not copied into memory in
    /// order to be rejected afterwards.
    private static let defaultMax = 16 * 1024 * 1024

    private var pending: FlutterResult?
    private var limit = FilePicker.defaultMax
    private weak var host: UIViewController?

    init(host: UIViewController?) {
        self.host = host
    }

    private func pick(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if pending != nil {
            result(FlutterError(code: "busy", message: "a picker is already open", details: nil))
            return
        }

        let args = call.arguments as? [String: Any]
        limit = args?["maxBytes"] as? Int ?? FilePicker.defaultMax
        pending = result

        if args?["images"] as? Bool == true, #available(iOS 14.0, *) {
            presentPhotos()
            return
        }

        let picker: UIDocumentPickerViewController
        if #available(iOS 14.0, *) {
            picker = UIDocumentPickerViewController(forOpeningContentTypes: [.item])
        } else {
            picker = UIDocumentPickerViewController(documentTypes: ["public.item"],
                                                    in: .import)
        }
        picker.delegate = self
        picker.allowsMultipleSelection = false

        guard let host = host else {
            pending = nil
            result(FlutterError(code: "nopicker", message: "no window to present from", details: nil))
            return
        }
        host.present(picker, animated: true)
    }

    func documentPicker(_ controller: UIDocumentPickerViewController,
                        didPickDocumentsAt urls: [URL]) {
        guard let waiting = pending else { return }
        pending = nil

        guard let url = urls.first else {
            result(waiting, nil)
            return
        }

        // Security scoped, and released whatever happens below.
        let scoped = url.startAccessingSecurityScopedResource()
        defer { if scoped { url.stopAccessingSecurityScopedResource() } }

        do {
            let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
            if let size = attributes[.size] as? Int, size > limit {
                waiting(FlutterError(code: "toolarge",
                                     message: "that file is \(size / 1024 / 1024) MB, and the limit is \(limit / 1024 / 1024) MB",
                                     details: nil))
                return
            }

            let data = try Data(contentsOf: url)
            if data.count > limit {
                waiting(FlutterError(code: "toolarge",
                                     message: "that file is larger than \(limit / 1024 / 1024) MB",
                                     details: nil))
                return
            }

            waiting([
                "name": url.lastPathComponent,
                "mime": mime(for: url),
                "bytes": FlutterStandardTypedData(bytes: data),
            ])
        } catch {
            waiting(FlutterError(code: "unreadable",
                                 message: error.localizedDescription,
                                 details: nil))
        }
    }

    func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        // Backed out. Null rather than an error: choosing nothing is a thing
        // people do and should not produce a message.
        let waiting = pending
        pending = nil
        waiting?(nil)
    }

    private func result(_ waiting: FlutterResult, _ value: Any?) {
        waiting(value)
    }

    private func mime(for url: URL) -> String {
        if #available(iOS 14.0, *),
           let type = UTType(filenameExtension: url.pathExtension),
           let mime = type.preferredMIMEType {
            return mime
        }
        return "application/octet-stream"
    }

    /// The system photograph grid.
    ///
    /// Out of process, so this application never sees the library, only the one
    /// picture that was chosen. Nothing is asked of the person and nothing
    /// appears on the privacy screen.
    @available(iOS 14.0, *)
    private func presentPhotos() {
        var config = PHPickerConfiguration()
        config.filter = .images
        config.selectionLimit = 1

        let picker = PHPickerViewController(configuration: config)
        picker.delegate = self

        guard let host = host else {
            let waiting = pending
            pending = nil
            waiting?(FlutterError(code: "nopicker",
                                  message: "no window to present from", details: nil))
            return
        }
        host.present(picker, animated: true)
    }

    func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "pick": pick(call, result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}

@available(iOS 14.0, *)
extension FilePicker: PHPickerViewControllerDelegate {

    func picker(_ picker: PHPickerViewController,
                didFinishPicking results: [PHPickerResult]) {
        picker.dismiss(animated: true)

        guard let waiting = pending else { return }
        pending = nil

        // Backed out. Null rather than an error, as with the other picker.
        guard let item = results.first?.itemProvider else {
            waiting(nil)
            return
        }

        // Asked for by type rather than as an image.
        //
        // `loadObject(ofClass: UIImage.self)` hands back a decoded bitmap, and
        // re-encoding it here would throw away whatever the camera wrote and
        // replace it with something larger. The bytes on disk are what is
        // wanted, and the size check below is about those bytes rather than
        // about a picture this process happened to redraw.
        let wanted = [UTType.jpeg, UTType.png, UTType.heic, UTType.gif]
        let type = wanted.first { item.hasItemConformingToTypeIdentifier($0.identifier) }
            ?? UTType.image

        item.loadDataRepresentation(forTypeIdentifier: type.identifier) { data, error in
            DispatchQueue.main.async {
                guard let data = data else {
                    waiting(FlutterError(
                        code: "unreadable",
                        message: error?.localizedDescription ?? "that picture could not be read",
                        details: nil))
                    return
                }

                if data.count > self.limit {
                    waiting(FlutterError(
                        code: "toolarge",
                        message: "that picture is \(data.count / 1024 / 1024) MB, "
                            + "and the limit is \(self.limit / 1024 / 1024) MB",
                        details: nil))
                    return
                }

                let name = item.suggestedName ?? "picture"
                let extension_ = type.preferredFilenameExtension ?? "jpg"

                waiting([
                    "name": name.contains(".") ? name : "\(name).\(extension_)",
                    "mime": type.preferredMIMEType ?? "image/jpeg",
                    "bytes": FlutterStandardTypedData(bytes: data),
                ])
            }
        }
    }
}

File: android/app/src/main/kotlin/com/rotelyx/app/FilePicker.kt
package com.ideoalabs.rotelyx

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.provider.OpenableColumns
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream

/**
 * Choosing a file, through the system picker and nothing else.
 *
 * # Why `ACTION_OPEN_DOCUMENT` and not a permission
 *
 * The obvious way to let somebody attach a photograph is to ask for
 * `READ_MEDIA_IMAGES` and browse their gallery. This does not, and the reason
 * is on the permission screen: an application that asks to read every image on
 * a phone has asked for something far larger than "let me send this one", and a
 * person reading that list has no way to tell the two apart.
 *
 * The Storage Access Framework asks for nothing. The system draws the picker,
 * the user chooses one file, and this process is handed a URI it may read
 * exactly once. Everything else on the device stays invisible to it, which is
 * the same shape as the rest of this application: the narrowest thing that
 * does the job.
 *
 * # Why the bytes are copied rather than the URI passed along
 *
 * A URI granted this way is valid for this task and not afterwards. Handing it
 * to Dart to open later produces a `SecurityException` at whatever moment the
 * user finally presses send, which is the worst place to discover it. So the
 * file is read here, while the grant is certainly alive, and what crosses the
 * channel is bytes.
 *
 * The ceiling matters for the same reason the mailbox has one: an attachment
 * becomes an envelope, envelopes are padded to a uniform size so that none
 * stands out, and something enormous cannot be padded into the crowd.
 */
class FilePicker(private val activity: Activity) {

    companion object {
        const val CHANNEL = "rotelyx/files"
        const val REQUEST = 0x51F0

        /** Refused before it is read, so a huge file is not copied into memory
         *  in order to be rejected afterwards. */
        private const val DEFAULT_MAX = 16 * 1024 * 1024
    }

    private var pending: MethodChannel.Result? = null
    private var limit = DEFAULT_MAX

    fun pick(call: MethodCall, result: MethodChannel.Result) {
        if (pending != null) {
            result.error("busy", "a picker is already open", null)
            return
        }

        limit = call.argument<Int>("maxBytes") ?: DEFAULT_MAX
        pending = result

        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "*/*"
        }

        try {
            activity.startActivityForResult(intent, REQUEST)
        } catch (e: Exception) {
            pending = null
            result.error("nopicker", "this device has no file picker", null)
        }
    }

    /** Returns whether this was ours to handle. */
    fun onResult(request: Int, code: Int, data: Intent?): Boolean {
        if (request != REQUEST) return false

        val waiting = pending ?: return true
        pending = null

        if (code != Activity.RESULT_OK) {
            // Backed out. Not an error: null means "chose nothing", which is a
            // thing people do and should not produce a message.
            waiting.success(null)
            return true
        }

        val uri = data?.data
        if (uri == null) {
            waiting.success(null)
            return true
        }

        try {
            waiting.success(read(uri))
        } catch (e: TooLarge) {
            waiting.error("toolarge", e.message, null)
        } catch (e: Exception) {
            waiting.error("unreadable", e.message ?: "that file could not be read", null)
        }
        return true
    }

    private class TooLarge(message: String) : Exception(message)

    private fun read(uri: Uri): Map<String, Any> {
        val resolver = activity.contentResolver

        var name = "attachment"
        var declared = -1L
        resolver.query(uri, null, null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) {
                val nameAt = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                if (nameAt >= 0 && !cursor.isNull(nameAt)) name = cursor.getString(nameAt)
                val sizeAt = cursor.getColumnIndex(OpenableColumns.SIZE)
                if (sizeAt >= 0 && !cursor.isNull(sizeAt)) declared = cursor.getLong(sizeAt)
            }
        }

        // Checked before reading, when the provider says how big it is. Some
        // providers do not, which is why the read below also counts.
        if (declared > limit) {
            throw TooLarge("that file is ${declared / 1024 / 1024} MB, and the limit is ${limit / 1024 / 1024} MB")
        }

        val out = ByteArrayOutputStream()
        resolver.openInputStream(uri).use { input ->
            if (input == null) throw Exception("that file could not be opened")
            val buffer = ByteArray(64 * 1024)
            while (true) {
                val read = input.read(buffer)
                if (read <= 0) break
                out.write(buffer, 0, read)
                if (out.size() > limit) {
                    throw TooLarge("that file is larger than ${limit / 1024 / 1024} MB")
                }
            }
        }

        return mapOf(
            "name" to name,
            "mime" to (resolver.getType(uri) ?: "application/octet-stream"),
            "bytes" to out.toByteArray(),
        )
    }

    fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "pick" -> pick(call, result)
            else -> result.notImplemented()
        }
    }
}
